diff --git a/src/samping/driver/sqs.py b/src/samping/driver/sqs.py
index 2c5e60e..22b6621 100644
--- a/src/samping/driver/sqs.py
+++ b/src/samping/driver/sqs.py
@@ -101,8 +101,6 @@ class SQSDriver(QueueDriver):
     async def _send_batch(self, queue: str, messages: List[Message]):
         async with self.sqs_client() as sqs:
             sqs_queue = await self.get_queue(sqs, queue)
-            print(queue)
-            print(sqs_queue.url)
             async with BatchWriter(sqs_queue) as batch:
                 for id, message in enumerate(messages):
                     body = base64.standard_b64encode(message.body).decode()
